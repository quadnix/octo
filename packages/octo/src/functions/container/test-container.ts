import type { Constructable } from '../../app.type.js';
import { DiffUtility } from '../diff/diff.utility.js';
import { Container } from './container.js';

type Factory<T> = { create: () => Promise<T> };

type FactoryMock<T> = {
  metadata?: { [key: string]: string };
  type: Constructable<T> | string;
  value: T;
};

type TestContainerSubjects = {
  importFrom?: Constructable<IPackageMock>[];
  mocks?: FactoryMock<unknown>[];
};
type TestContainerOptions = { factoryTimeoutInMs?: number };

export interface IPackageMock {
  destroy: () => Promise<void>;

  getMocks: () => FactoryMock<unknown>[];

  init(): Promise<void>;
}

/**
 * The TestContainer class is an isolated {@link Container} for tests.
 *
 * :::info Info
 * A test framework like Jest, supports parallel test executions in different cores of the machine.
 * Each test is a single test file, whose `it()` blocks are executed serially.
 * :::
 *
 * TestContainer is created with testing framework's parallel execution in mind.
 * A TestContainer should be created in the `beforeAll()` block of your test,
 * which modifies the Container for the duration of the test.
 * This TestContainer does not affect other tests in parallel since they are on a separate machine core.
 * Once tests are done executing, the `afterAll()` block cleans up the Container.
 */
export class TestContainer {
  private static packageMocks: IPackageMock[] = [];

  private static createTestFactory<T>(mock: FactoryMock<T>): Factory<T> {
    return class {
      static mock: FactoryMock<T> = mock;

      static async create(): Promise<T> {
        return this.mock.value;
      }
    };
  }

  /**
   * The `TestContainer.create()` method allows you to mock factories.
   *
   * @example
   * ```ts
   * beforeAll(async () => {
   *   await TestContainer.create({
   *     importFrom: [OctoAwsCdkPackageMock],
   *     mocks: [
   *       { type: MyClass, value: jest.fn() },
   *       { metadata: { key: 'value' }, type: AnotherClass, value: new AnotherClass() },
   *     ],
   *   }, { factoryTimeoutInMs: 500 });
   * });
   *
   * ```
   * @param subjects The subjects being mocked.
   * - `importFrom` is an array of {@link IPackageMock} classes, to import custom mocks,
   * such as from octo-aws-cdk package - [OctoAwsCdkPackageMock](/api/octo-aws-cdk/class/OctoAwsCdkPackageMock).
   * - `mocks` is an array of objects, to override the default factories.
   *   - Use `metadata?: { [key: string]: string }` to identify the factory being mocked.
   *   - Use `type: Constructable<T> | string` to identify the class being mocked.
   *   - Use `value: T` to provide the mocked value.
   * @param options Options to configure TestContainer.
   * - `factoryTimeoutInMs?: number` is to override the default container timeout.
   */
  static async create(subjects: TestContainerSubjects, options?: TestContainerOptions): Promise<void> {
    const oldFactories = { ...Container['factories'] };
    Container.reset();

    if (options?.factoryTimeoutInMs) {
      Container.setFactoryTimeout(options.factoryTimeoutInMs);
    }

    const importedMocks: FactoryMock<unknown>[] = [];
    for (const mockClass of subjects.importFrom || []) {
      const mock = new mockClass();
      await mock.init();

      importedMocks.push(...mock.getMocks());
      this.packageMocks.push(mock);
    }

    subjects.mocks = [...importedMocks, ...(subjects.mocks || [])];

    for (const mock of subjects.mocks) {
      const mockClassName = typeof mock.type === 'string' ? mock.type : mock.type.name;
      const mockMetadata = mock.metadata || {};
      const registeredFactories = oldFactories[mockClassName] || [];
      const matchingFactoryIndex = registeredFactories.findIndex((f) =>
        DiffUtility.isObjectDeepEquals(f.metadata, mockMetadata),
      );

      if (matchingFactoryIndex > -1) {
        oldFactories[mockClassName][matchingFactoryIndex].factory = this.createTestFactory(mock);
      }
    }

    for (const name in oldFactories) {
      for (const oldFactory of oldFactories[name]) {
        Container.registerFactory(name, oldFactory.factory as Factory<unknown>, { metadata: oldFactory.metadata });
        if (oldFactory.default) {
          Container.setDefault(name, oldFactory.factory as Factory<unknown>);
        }
      }
    }
  }

  /**
   * The `TestContainer.reset()` method will completely destroy all factories from the Container.
   * Because the entire Container is destroyed, and not just the mocks,
   * it must always be called in your `afterAll()` block.
   *
   * @example
   * ```ts
   * afterAll(async () => {
   *   await TestContainer.reset();
   * });
   * ```
   */
  static async reset(): Promise<void> {
    for (const mock of this.packageMocks) {
      await mock.destroy();
    }
    Container.reset();
  }
}
