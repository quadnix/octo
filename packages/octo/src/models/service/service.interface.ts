import { Service } from './service.model.js';

export interface IService {
  serviceId: Service['serviceId'];
}
