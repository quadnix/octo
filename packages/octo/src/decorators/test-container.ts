import { Constructable } from '../app.type.js';
import { DiffUtility } from '../functions/diff/diff.utility.js';
import { Container } from './container.js';

type Factory<T> = { create: () => Promise<T> };

type FactoryMock<T> = {
  metadata?: { [key: string]: string };
  type: Constructable<T> | string;
  value: T;
};

type TestContainerOptions = { factoryTimeoutInMs?: number };

export class TestContainer {
  private static createTestFactory<T>(mock: FactoryMock<T>): Factory<T> {
    return class {
      static mock: FactoryMock<T> = mock;

      static async create(): Promise<T> {
        return this.mock.value;
      }
    };
  }

  static create(mocks: FactoryMock<unknown>[], options: TestContainerOptions): void {
    const oldFactories = { ...Container['factories'] };
    Container.reset();

    if (options.factoryTimeoutInMs) {
      Container.setFactoryTimeout(options.factoryTimeoutInMs);
    }

    for (const mock of mocks) {
      const mockClassName = typeof mock.type === 'string' ? mock.type : mock.type.name;
      const mockMetadata = mock.metadata || {};
      const registeredFactories = oldFactories[mockClassName] || [];
      const matchingFactoryIndex = registeredFactories.findIndex((f) =>
        DiffUtility.isObjectDeepEquals(f.metadata, mockMetadata),
      );

      if (matchingFactoryIndex > -1) {
        oldFactories[mockClassName][matchingFactoryIndex].factory = this.createTestFactory(mock);
      }
    }

    for (const name in oldFactories) {
      for (const oldFactory of oldFactories[name]) {
        Container.registerFactory(name, oldFactory.factory as Factory<unknown>, { metadata: oldFactory.metadata });
        if (oldFactory.default) {
          Container.setDefault(name, oldFactory.factory as Factory<unknown>);
        }
      }
    }
  }
}
