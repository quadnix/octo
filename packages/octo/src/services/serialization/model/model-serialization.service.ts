import {
  type IUnknownModel,
  type ModelSerializedOutput,
  NodeType,
  type UnknownModel,
  type UnknownOverlay,
} from '../../../app.type.js';
import { Container } from '../../../functions/container/container.js';
import { EventSource } from '../../../decorators/event-source.decorator.js';
import { Factory } from '../../../decorators/factory.decorator.js';
import {
  AnchorRegistrationEvent,
  ModelDeserializedEvent,
  ModelRegistrationEvent,
  ModelSerializedEvent,
  OverlayRegistrationEvent,
} from '../../../events/index.js';
import { Dependency, type IDependency } from '../../../functions/dependency/dependency.js';
import type { ANode } from '../../../functions/node/node.abstract.js';
import type { AAnchor } from '../../../overlays/anchor.abstract.js';
import type { IAnchor } from '../../../overlays/anchor.interface.js';
import { OverlayDataRepository } from '../../../overlays/overlay-data.repository.js';
import type { IOverlay } from '../../../overlays/overlay.interface.js';
import { ObjectUtility } from '../../../utilities/object/object.utility.js';

export class ModelSerializationService {
  private MODEL_DESERIALIZATION_TIMEOUT_IN_MS = 5000;

  private readonly classMapping: { [key: string]: any } = {};

  private async _deserialize(
    serializedOutput: ModelSerializedOutput,
    freeze: boolean,
  ): Promise<{ overlays: UnknownOverlay[]; root: UnknownModel }> {
    const deReferencePromises: {
      [p: string]: [Promise<boolean>, (value: boolean) => void, (error: Error) => void, NodeJS.Timeout];
    } = {};
    const seen: { [p: string]: UnknownModel } = {};

    const deReferenceContext = async (context: string): Promise<UnknownModel> => {
      if (!seen[context]) {
        if (deReferencePromises[context]) {
          await deReferencePromises[context][0];
        } else {
          deReferencePromises[context] = [] as any;
          const promise = new Promise<boolean>((resolve, reject) => {
            deReferencePromises[context][1] = resolve;
            deReferencePromises[context][2] = reject;
          });
          deReferencePromises[context][0] = promise;

          deReferencePromises[context][3] = setTimeout(() => {
            deReferencePromises[context][2](new Error('DeReferencing model operation timed out!'));
          }, this.MODEL_DESERIALIZATION_TIMEOUT_IN_MS);
          await promise;
        }
      }

      return seen[context];
    };

    const deserializeModel = async (context: string): Promise<UnknownModel> => {
      const { className, model } = serializedOutput.models[context];
      const deserializationClass = this.classMapping[className];

      seen[context] = await deserializationClass.unSynth(model, deReferenceContext);
      seen[context]['context'] = context;

      if (deReferencePromises[context] !== undefined) {
        deReferencePromises[context][1](true);
        clearTimeout(deReferencePromises[context][3]);
      }

      return seen[context];
    };

    // Deserialize all models.
    const promiseToDeserializeModels: Promise<UnknownModel>[] = [];
    for (const context in serializedOutput.models) {
      promiseToDeserializeModels.push(deserializeModel(context));
    }
    await Promise.all(promiseToDeserializeModels);

    // Deserialize all anchors.
    for (const a of serializedOutput.anchors) {
      const { className, parent } = a;
      const deserializationClass = this.classMapping[className];

      const anchor = await deserializationClass.unSynth(deserializationClass, a, deReferenceContext);
      seen[parent.context].addAnchor(anchor);
    }

    // Deserialize all overlays.
    const overlays: UnknownOverlay[] = [];
    for (const { className, overlay } of serializedOutput.overlays) {
      const deserializationClass = this.classMapping[className];
      const newOverlay = await deserializationClass.unSynth(deserializationClass, overlay, deReferenceContext);
      overlays.push(newOverlay);
      seen[newOverlay.getContext()] = newOverlay;
    }

    // Deserialize all dependencies.
    for (const d of serializedOutput.dependencies) {
      const dependency = Dependency.unSynth(seen[d.from], seen[d.to], d);
      if (!seen[d.from].getDependencies().some((d) => d.from === dependency.from && d.to === dependency.to)) {
        seen[d.from]['dependencies'].push(dependency);
      }
    }

    if (freeze) {
      for (const model of Object.values(seen)) {
        ObjectUtility.deepFreeze(model);
      }
      for (const overlay of overlays) {
        ObjectUtility.deepFreeze(overlay);
      }
    }

    // If no dependencies to serialize, return the first seen model.
    const root =
      serializedOutput.dependencies.length > 0
        ? seen[serializedOutput.dependencies[0].from]
        : seen[Object.keys(seen)[0]];

    return { overlays, root };
  }

  @EventSource(ModelDeserializedEvent)
  async deserialize(
    serializedOutput: ModelSerializedOutput,
    { freeze = true }: { freeze?: boolean } = {},
  ): Promise<UnknownModel> {
    const { overlays: oldOverlays, root } = await this._deserialize(serializedOutput, freeze);

    // Refresh the overlay data repository.
    await Container.getInstance().get(OverlayDataRepository, { args: [true, oldOverlays] });

    return root;
  }

  @EventSource(AnchorRegistrationEvent)
  @EventSource(ModelRegistrationEvent)
  @EventSource(OverlayRegistrationEvent)
  registerClass(className: string, deserializationClass: any): void {
    if (this.classMapping[className]) {
      throw new Error(`Class "${className}" is already registered!`);
    }
    this.classMapping[className] = deserializationClass;
  }

  @EventSource(ModelSerializedEvent)
  async serialize(root: UnknownModel): Promise<ModelSerializedOutput> {
    const boundary = root.getBoundaryMembers();
    const anchors: (IAnchor & { className: string })[] = [];
    const dependencies: IDependency[] = [];
    const models: { [key: string]: { className: string; model: IUnknownModel } } = {};
    const overlays: { className: string; overlay: IOverlay }[] = [];

    for (const model of boundary) {
      for (const d of model.getDependencies()) {
        // Skip dependencies that are not part of boundary.
        if (
          !boundary.some((m) => m.getContext() === d.from.getContext()) ||
          !boundary.some((m) => m.getContext() === d.to.getContext())
        ) {
          continue;
        }

        dependencies.push(d.synth());
      }

      if ((model.constructor as typeof ANode).NODE_TYPE === NodeType.MODEL) {
        const context = model.getContext();
        if (!models[context]) {
          models[context] = {
            className: `${(model.constructor as typeof ANode).NODE_PACKAGE}/${model.constructor.name}`,
            model: model.synth() as IUnknownModel,
          };
        }

        for (const a of (model as UnknownModel).getAnchors()) {
          anchors.push({
            ...a.synth(),
            className: `${(a.constructor as typeof AAnchor).NODE_PACKAGE}/${a.constructor.name}`,
          });
        }
      } else if ((model.constructor as typeof ANode).NODE_TYPE === NodeType.OVERLAY) {
        overlays.push({
          className: `${(model.constructor as typeof ANode).NODE_PACKAGE}/${model.constructor.name}`,
          overlay: (model as UnknownOverlay).synth(),
        });
      }
    }

    return { anchors, dependencies, models, overlays };
  }

  setModelDeserializationTimeout(timeoutInMs: number): void {
    this.MODEL_DESERIALIZATION_TIMEOUT_IN_MS = timeoutInMs;
  }
}

@Factory<ModelSerializationService>(ModelSerializationService)
export class ModelSerializationServiceFactory {
  private static instance: ModelSerializationService;

  static async create(forceNew = false): Promise<ModelSerializationService> {
    if (!this.instance) {
      this.instance = new ModelSerializationService();
    }
    if (forceNew) {
      const newInstance = new ModelSerializationService();
      Object.keys(this.instance).forEach((key) => (this.instance[key] = newInstance[key]));
    }
    return this.instance;
  }
}
