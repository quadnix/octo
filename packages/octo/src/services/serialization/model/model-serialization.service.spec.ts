import { TestAnchor, TestOverlay } from '../../../utilities/test-helpers/test-classes.js';
import { commit, create, createTestOverlays } from '../../../utilities/test-helpers/test-models.js';
import { type ModelSerializedOutput } from '../../../app.type.js';
import type { Container } from '../../../functions/container/container.js';
import { TestContainer } from '../../../functions/container/test-container.js';
import { type IDependency } from '../../../functions/dependency/dependency.js';
import { Account } from '../../../models/account/account.model.js';
import { App } from '../../../models/app/app.model.js';
import { Image } from '../../../models/image/image.model.js';
import { Region } from '../../../models/region/region.model.js';
import { Service } from '../../../models/service/service.model.js';
import { Subnet } from '../../../models/subnet/subnet.model.js';
import { OverlayDataRepository } from '../../../overlays/overlay-data.repository.js';
import { ModelSerializationService } from './model-serialization.service.js';

describe('Model Serialization Service UT', () => {
  let container: Container;

  beforeEach(async () => {
    container = await TestContainer.create({ mocks: [] }, { factoryTimeoutInMs: 500 });

    const modelSerializationService = new ModelSerializationService();
    modelSerializationService.registerClass('@octo/Account', Account);
    modelSerializationService.registerClass('@octo/App', App);
    modelSerializationService.registerClass('@octo/Image', Image);
    modelSerializationService.registerClass('@octo/Region', Region);
    modelSerializationService.registerClass('@octo/Service', Service);
    modelSerializationService.registerClass('@octo/Subnet', Subnet);
    modelSerializationService.registerClass('@octo/TestAnchor', TestAnchor);
    modelSerializationService.registerClass('@octo/TestOverlay', TestOverlay);
    container.unRegisterFactory(ModelSerializationService);
    container.registerValue(ModelSerializationService, modelSerializationService);
  });

  afterEach(async () => {
    await TestContainer.reset();
  });

  describe('deserialize()', () => {
    it('should throw error when de-serializing an unknown class', async () => {
      const service = await container.get(ModelSerializationService);

      const serializedOutput: ModelSerializedOutput = {
        anchors: [],
        dependencies: [
          {
            from: 'app=name',
          } as IDependency,
        ],
        models: {
          'app=name': { className: 'ClassNotExist', model: null },
        } as any,
        overlays: [],
      };

      await expect(async () => {
        await service.deserialize(serializedOutput);
      }).rejects.toThrow();
    });

    it('should throw error when de-serializing a class with default unSynth', async () => {
      const service = await container.get(ModelSerializationService);

      const serializedOutput: ModelSerializedOutput = {
        anchors: [],
        dependencies: [
          {
            from: 'app=name',
          } as IDependency,
        ],
        models: {
          'app=name': { className: '@octo/Service', model: null },
        } as any,
        overlays: [],
      };

      await expect(async () => {
        await service.deserialize(serializedOutput);
      }).rejects.toThrowErrorMatchingInlineSnapshot(`"Method not implemented! Use derived class implementation"`);
    });

    it('should be able to register classes and deserialize', async () => {
      const {
        app: [app],
      } = create({ account: ['aws,account'], app: ['test-app'] });

      const app_1 = await commit(app);

      expect(app_1.name).toBe('test-app');
    });

    it('should deserialize a single model', async () => {
      const {
        app: [app],
      } = create({ app: ['test-app'] });

      const app_1 = await commit(app);

      expect(app_1.name).toBe('test-app');
    });

    it('should return the serialized root on deserialization', async () => {
      const {
        region: [region],
      } = create({ account: ['aws,account'], app: ['test-app'], region: ['region'] });

      const region_1 = await commit(region);

      expect(region_1.regionId).toBe('region');
    });

    it('should have exact same dependencies on deserialized object as original object', async () => {
      const service = await container.get(ModelSerializationService);

      const {
        app: [app],
      } = create({
        account: ['aws,account'],
        app: ['test'],
        image: ['image'],
        region: ['region'],
        subnet: ['private', 'public:-1'],
      });

      const app_1 = await commit(app);

      const previousDependencies = (await service.serialize(app_1)).dependencies.sort((a, b) =>
        a.from + a.to > b.from + b.to ? 1 : b.from + b.to > a.from + a.to ? -1 : 0,
      );
      const currentDependencies = (await service.serialize(app)).dependencies.sort((a, b) =>
        a.from + a.to > b.from + b.to ? 1 : b.from + b.to > a.from + a.to ? -1 : 0,
      );
      expect(previousDependencies).toEqual(currentDependencies);
    });

    it('should deserialize overlay with multiple anchors of same parent', async () => {
      const overlayDataRepository = await container.get(OverlayDataRepository);

      const {
        app: [app],
      } = create({ account: ['aws,account'], app: ['test-app'] });
      const anchor1 = new TestAnchor('anchor-1', {}, app);
      app.addAnchor(anchor1);
      const anchor2 = new TestAnchor('anchor-2', {}, app);
      app.addAnchor(anchor2);

      await createTestOverlays({ 'overlay-1': [anchor1, anchor2] });

      const app_1 = await commit(app);
      expect(overlayDataRepository['newOverlays']).toHaveLength(1);

      const overlay1_1 = overlayDataRepository['newOverlays'][0];
      expect(overlay1_1.getAnchors().map((a) => a.getParent().getContext())).toEqual([
        app_1.getContext(),
        app_1.getContext(),
      ]);
    });

    /**
     * Users are not required to deserialize models and overlays because models and overlays are not diff-ed.
     * This feature is only kept around for future use by Octo web clients.
     * Upon deserialization, a new copy of models are created, and the old overlays are kept in OverlayDataRepository
     * as new overlays. Even the web client would only ever deserialize models and overlays to programmatically
     * read the old state.
     */
    it('should deserialize overlay and ensure it still points to the old models', async () => {
      const overlayDataRepository = await container.get(OverlayDataRepository);

      const {
        account: [account],
        app: [app],
      } = create({ account: ['aws,account'], app: ['test-app'], image: ['image'] });
      const anchor1 = new TestAnchor('anchor-1', {}, app);
      app.addAnchor(anchor1);

      await createTestOverlays({ 'overlay-1': [anchor1] });

      // Before commit the overlay has reference to the app, and the app has no region children.
      expect(overlayDataRepository['newOverlays'][0].getAnchors()[0].getParent().getChildren()['region']).toBe(
        undefined,
      );
      await commit(app);

      // Modify the new app.
      account.addRegion(new Region('region-1'));

      // After commit the overlay is deserialized and the newOverlays still reference the old models.
      // Updates to new models should not be reflected in the old models.
      expect(overlayDataRepository['newOverlays'][0].getAnchors()[0].getParent().getChildren()['region']).toBe(
        undefined,
      );
    });
  });

  describe('serialize()', () => {
    it('should serialize an empty app', async () => {
      const {
        app: [app],
      } = create({ app: ['test-app'] });

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(app)).toMatchSnapshot();
    });

    it('should serialize a non-empty app', async () => {
      const {
        app: [app],
      } = create({
        account: ['aws,account'],
        app: ['test-app'],
        environment: ['qa'],
        image: ['image'],
        region: ['region-1'],
        server: ['backend'],
      });

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(app)).toMatchSnapshot();
    });

    it('should serialize model anchors', async () => {
      const {
        app: [app],
      } = create({ app: ['test-app'] });
      const anchor1 = new TestAnchor('anchor-1', {}, app);
      app.addAnchor(anchor1);

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(app)).toMatchSnapshot();
    });

    it('should serialize only boundary members', async () => {
      const {
        region: [region1],
      } = create({
        account: ['aws,account'],
        app: ['test-app'],
        environment: ['qa', 'qa'],
        region: ['region-1', 'region-2:-1'],
      });

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(region1)).toMatchSnapshot();
    });

    it('should serialize overlay with multiple anchors of same parent', async () => {
      const {
        app: [app],
      } = create({ account: ['aws,account'], app: ['test-app'] });
      const anchor1 = new TestAnchor('anchor-1', {}, app);
      app.addAnchor(anchor1);
      const anchor2 = new TestAnchor('anchor-2', {}, app);
      app.addAnchor(anchor2);

      await createTestOverlays({ 'overlay-1': [anchor1, anchor2] });

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(app)).toMatchSnapshot();
    });

    it('should serialize two overlay dependencies with each other', async () => {
      const {
        app: [app],
      } = create({ account: ['aws,account'], app: ['test-app'] });
      const anchor1 = new TestAnchor('anchor-1', {}, app);
      app.addAnchor(anchor1);
      const anchor2 = new TestAnchor('anchor-2', {}, app);
      app.addAnchor(anchor2);

      const [overlay1, overlay2] = await createTestOverlays({ 'overlay-1': [anchor1], 'overlay-2': [anchor2] });
      overlay1.addRelationship(overlay2);

      const service = await container.get(ModelSerializationService);

      expect(await service.serialize(app)).toMatchSnapshot();
    });
  });
});
