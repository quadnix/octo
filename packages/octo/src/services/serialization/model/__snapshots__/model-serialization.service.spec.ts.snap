// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model Serialization Service UT serialize() should serialize a non-empty app 1`] = `
{
  "anchors": [],
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "imageId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "imageId",
        "type": "parent",
      },
      "to": "image=image:0.0.1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "regionId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "regionId",
        "type": "parent",
      },
      "to": "region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "serverKey",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "serverKey",
        "type": "parent",
      },
      "to": "server=backend,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "serverKey",
          "toField": "name",
        },
      ],
      "from": "server=backend,app=test-app",
      "relationship": {
        "onField": "serverKey",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "environmentName",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "environmentName",
        "type": "parent",
      },
      "to": "environment=qa,region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
      ],
      "from": "environment=qa,region=region-1,app=test-app",
      "relationship": {
        "onField": "environmentName",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "imageId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "imageId",
          "toField": "name",
        },
      ],
      "from": "image=image:0.0.1,app=test-app",
      "relationship": {
        "onField": "imageId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "environment=qa,region=region-1,app=test-app": {
      "className": "Environment",
      "model": {
        "environmentName": "qa",
        "environmentVariables": {
          "key": "value",
        },
      },
    },
    "image=image:0.0.1,app=test-app": {
      "className": "Image",
      "model": {
        "dockerOptions": {
          "buildArgs": {},
          "dockerfilePath": "/Dockerfile",
          "quiet": undefined,
        },
        "imageId": "0.0.1",
        "imageName": "image",
        "imageTag": "0.0.1",
      },
    },
    "region=region-1,app=test-app": {
      "className": "Region",
      "model": {
        "regionId": "region-1",
      },
    },
    "server=backend,app=test-app": {
      "className": "Server",
      "model": {
        "serverKey": "backend",
      },
    },
  },
  "overlays": [],
}
`;

exports[`Model Serialization Service UT serialize() should serialize an empty app 1`] = `
{
  "anchors": [],
  "dependencies": [],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
  },
  "overlays": [],
}
`;

exports[`Model Serialization Service UT serialize() should serialize only boundary members 1`] = `
{
  "anchors": [],
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "environmentName",
        },
      ],
      "from": "region=region-1,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "environmentName",
        "type": "parent",
      },
      "to": "environment=qa,region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "regionId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "regionId",
        "type": "parent",
      },
      "to": "region=region-1,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "environmentName",
          "toField": "regionId",
        },
      ],
      "from": "environment=qa,region=region-1,app=test-app",
      "relationship": {
        "onField": "environmentName",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region-1,app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "environment=qa,region=region-1,app=test-app": {
      "className": "Environment",
      "model": {
        "environmentName": "qa",
        "environmentVariables": {},
      },
    },
    "region=region-1,app=test-app": {
      "className": "Region",
      "model": {
        "regionId": "region-1",
      },
    },
  },
  "overlays": [],
}
`;

exports[`Model Serialization Service UT serialize() should serialize when multiple models have dependency on same model 1`] = `
{
  "anchors": [],
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "name",
          "toField": "regionId",
        },
      ],
      "from": "app=test-app",
      "relationship": {
        "onField": "name",
        "toField": "regionId",
        "type": "parent",
      },
      "to": "region=region,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "regionId",
          "toField": "name",
        },
      ],
      "from": "region=region,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "name",
        "type": "child",
      },
      "to": "app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "subnetId",
        },
      ],
      "from": "region=region,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "subnetId",
        "type": "parent",
      },
      "to": "subnet=region-private,region=region,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "regionId",
          "toField": "subnetId",
        },
      ],
      "from": "region=region,app=test-app",
      "relationship": {
        "onField": "regionId",
        "toField": "subnetId",
        "type": "parent",
      },
      "to": "subnet=region-public,region=region,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "subnetId",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "subnetId",
          "toField": "regionId",
        },
      ],
      "from": "subnet=region-public,region=region,app=test-app",
      "relationship": {
        "onField": "subnetId",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region,app=test-app",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "subnetId",
          "toField": "regionId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "subnetId",
          "toField": "regionId",
        },
      ],
      "from": "subnet=region-private,region=region,app=test-app",
      "relationship": {
        "onField": "subnetId",
        "toField": "regionId",
        "type": "child",
      },
      "to": "region=region,app=test-app",
    },
  ],
  "models": {
    "app=test-app": {
      "className": "App",
      "model": {
        "name": "test-app",
      },
    },
    "region=region,app=test-app": {
      "className": "Region",
      "model": {
        "regionId": "region",
      },
    },
    "subnet=region-private,region=region,app=test-app": {
      "className": "Subnet",
      "model": {
        "options": {
          "disableSubnetIntraNetwork": false,
          "subnetType": "private",
        },
        "region": {
          "context": "region=region,app=test-app",
        },
        "subnetId": "region-private",
        "subnetName": "private",
      },
    },
    "subnet=region-public,region=region,app=test-app": {
      "className": "Subnet",
      "model": {
        "options": {
          "disableSubnetIntraNetwork": false,
          "subnetType": "private",
        },
        "region": {
          "context": "region=region,app=test-app",
        },
        "subnetId": "region-public",
        "subnetName": "public",
      },
    },
  },
  "overlays": [],
}
`;
